### PR TITLE
Remove APIStatus from the client

### PR DESCRIPTION
**Pull request by krousey**
Everything that used it also depended on pkg/api/errors. It was a
needless abstraction since it was always an *errors.StatusError.

cc @kubernetes/sig-api-machinery @kubernetes/goog-csi

### CODE REVIEW

**Comment by smarterclayton** [comment]
This is not a needless abstraction - there can be other error types  Why is pointing to a struct directly better than pointing to an interface which can be clearly mocked and provided by other clients?

I'm not in favor of this change without more justification.

**Comment by krousey** [comment]
> there can be other error types

No other error type currently implements this interface. Any type that did would be incompatible with pkg/api/errors. Every use of this interface in the codebase coincides with use of pkg/api/errors. Only one test file even had to have the dependency added.

> Why is pointing to a struct directly better than pointing to an interface

Because the interface is completely redundant and unnecessary. It offers no abstraction benefits since all of the uses are already tied to pkg/api/errors and this struct is simple to use.

> which can be clearly mocked and provided by other clients?

Why would you? This struct is dead simple, just put a Status object in it. A mock is complete over kill. Another implementation is also overkill, and would also not work with the existing pkg/api/errors package.

> I'm not in favor of this change without more justification.

Other than just being unnecessary, my other justification for doing this was reducing dependencies. If such an interface is necessary (and I would want justification for that), then it should probably be in the pkg/api/errors package, be made to work with that package, and removed from the client.

**Comment by krousey** [comment]
@smarterclayton mind if I assign you as reviewer?

**Comment by smarterclayton** [comment]
In general, go defines interfaces where they are needed (in client package)
not in the source package.  Any client consumer has the ability to provide
errors that match the version status interface.  In addition, go typically
advocates keeping error types private unless there is structure data
available.  Neither of those are so strong though that they override the
"less is more argument", but they were chosen that way deliberately

The primary reason the interface is addressed is keep the concrete type
hidden in API/errors.  The secondary reason is to allow rest clients to
provide synthetic server errors - for instance, to allow a client Transport
wrapper to return a synthetic status error in certain error conditions
without having to allocate a status error immediately.  For instance, a
Kerberos transport wrapper may return an unauthorized status response
created synthetically, but only rest operations may use the additional
API.Status info (why a regular http.Client may not).  Not a total deal
breaker, but I don't like exposing error types in general (which is why we
use the idiomatic Is*Error() function to hide that detail).


On Nov 30, 2015, at 8:08 PM, krousey <notifications@github.com> wrote:

there can be other error types

No other error type currently implements this interface. Any type that did
would be incompatible with pkg/api/errors. Every use of this interface in
the codebase coincides with use of pkg/api/errors. Only one test file even
had to have the dependency added.

Why is pointing to a struct directly better than pointing to an interface

Because the interface is completely redundant and unnecessary. It offers no
abstraction benefits since all of the uses are already tied to
pkg/api/errors and this struct is simple to use.

which can be clearly mocked and provided by other clients?

Why would you? This struct is dead simple, just put a Status object in it.
A mock is complete over kill. Another implementation is also overkill, and
would also not work with the existing pkg/api/errors package.

I'm not in favor of this change without more justification.

Other than just being unnecessary, my other justification for doing this
was reducing dependencies. If such an interface is necessary (and I would
want justification for that), then it should probably be in the
pkg/api/errors package, be made to work with that package, and removed from
the client.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/17991#issuecomment-160814054>
.

**Comment by ncdc** [comment]
cc @kubernetes/rh-cluster-infra

**Comment by krousey** [comment]
@smarterclayton thoughts on how to resolve this? I don't think APIStatus should be a public type in the client package.

**Comment by smarterclayton** [comment]
Actually, I take back an earlier statement, it is used by consumers of
client.  We use it to check whether the returned error from the client has
structured data (like the invalid field data, or the exact code returned by
the server).  So a client consumer needs some abstraction for errors
created by other code that want to provide API response data, and that
abstraction today is APIStatus being implemented by errors.

On Dec 4, 2015, at 4:36 PM, krousey <notifications@github.com> wrote:

@smarterclayton <https://github.com/smarterclayton> thoughts on how to
resolve this? I don't think APIStatus should be a public type in the client
package.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/17991#issuecomment-162088056>
.

**Comment by lavalamp** [comment]
So I gather Kris's points are
1. This type doesn't reduce the import tree for client consumers
2. This type doesn't provide any useful abstraction for client consumers

Clayton's points are
1. This type allows structured error injection (e.g., by a custom transport)
2. This type allows things that aren't status errors but that have the ability to construct an unversioned.Status to still be passed as an error

My thoughts:
1. There needs to be some sort of Schelling point for introspectable structured client errors. I don't particularly care what it is. It probably should live in a subdirectory of the client package, e.g. `pkg/client/errors` or some such. It's fine by me if that currently is just a little blob that rehashes `pkg/api/errors` for the moment. We'd like users of the client library to not have to import pkg/api/..., because we'd like to ship server and client api specifications separately (eventually), so making clients manually `import "pkg/api/errors"` is not ideal.
2. The ultimate form of this interface should therefore probably not reference pkg/api/errors.
3. I don't know if I actually buy Clayton's second point as being practically useful (does anyone want to do that?); Does openshift even do that?

TL;DR maybe instead of removing this, move it to pkg/client/errors with a big TODO to eventually not depend on pkg/api/errors. Also consider just moving or duplicating the StatusError type-- having the client code reuse the server code is a little suspect.

**Comment by krousey** [comment]
Perhaps I should explain why I'm trying to change this now. Given the new client proposal, we're going to separate out RESTClient into it's own package and generate a bunch of clients on top of that. RESTclient currently uses APIStatus, but RESTclient can't depend on pkg/client/unversioned. We could leave APIStatus in pkg/client/unversioned and duplicate it in RESTClient. I think that would lead to confusion. So I went looking to see if there was a better home for APIStatus. Then I looked at pkg/api/errors. I think this is the best place for all types related to api.Status and errors. While I was reading that package, I noticed StatusError and I thought this was perfect. I could just delete APIStatus and change everything that was using it to use StatusError instead. I didn't think it would be a problem, since we already do this in many places:

 - [pkg/api/errors/errors.go#L419](https://github.com/kubernetes/kubernetes/blob/b7dc1175ba64ce36a52fa44910ea7dcb283ffed9/pkg/api/errors/errors.go#L419)
 - [pkg/api/errors/errors.go#L432](https://github.com/kubernetes/kubernetes/blob/b7dc1175ba64ce36a52fa44910ea7dcb283ffed9/pkg/api/errors/errors.go#L432)
 - [pkg/api/errors/errors_test.go#L154](https://github.com/kubernetes/kubernetes/blob/b7dc1175ba64ce36a52fa44910ea7dcb283ffed9/pkg/api/errors/errors_test.go#L154)
 - [pkg/client/record/event.go#L145](https://github.com/kubernetes/kubernetes/blob/b7dc1175ba64ce36a52fa44910ea7dcb283ffed9/pkg/client/record/event.go#L145)
 - [pkg/kubectl/cmd/util/helpers.go#L60](https://github.com/kubernetes/kubernetes/blob/b7dc1175ba64ce36a52fa44910ea7dcb283ffed9/pkg/kubectl/cmd/util/helpers.go#L60)
 - [pkg/registry/pod/etcd/etcd.go#L184](https://github.com/kubernetes/kubernetes/blob/b7dc1175ba64ce36a52fa44910ea7dcb283ffed9/pkg/registry/pod/etcd/etcd.go#L184)
 - [test/e2e/daemon_set.go#L238](https://github.com/kubernetes/kubernetes/blob/b7dc1175ba64ce36a52fa44910ea7dcb283ffed9/test/e2e/daemon_set.go#L238)

It's even what we document that we'll return here: [pkg/client/unversioned/request.go#L771](https://github.com/kubernetes/kubernetes/blob/b7dc1175ba64ce36a52fa44910ea7dcb283ffed9/pkg/client/unversioned/request.go#L771)

So I'm fine with disagreeing about the right approach regarding interface vs. struct. Whatever type we end up with, I strongly believe it should not be defined in the client. Here's a list of outcomes that I think would work for our goal of splitting off the RESTClient and generating versioned clients. I've ordered them is most to least preferential for me.

1. Remove the APIStatus interface and change all uses to use StatusError instead.
2. Move APIStatus to the errors package, change all references to use the new location, and update the errors package to type assert to the interface instead of the struct.
3. Move APIStatus to the new RESTclient package (in the same PR where it's created), and change all references to use the new location.
4. Duplicate the APIStatus to the RESTClient package (in the same PR where it's created).
5. ???

Is there one here (or one you could fill in for 5) that you'd be OK with?

**Comment by smarterclayton** [comment]
I'm happiest with 2.

Here's where we use it today:

*
https://github.com/openshift/origin/blob/master/pkg/cmd/cli/cmd/buildlogs.go#L44
*
https://github.com/openshift/origin/blob/master/pkg/cmd/cli/cmd/edit.go#L345

If StatusError had public members and could be constructed by any package,
I'd be ok with 1.  However, if you do that you're effectively opening
implementation details to all consumers, so I'd prefer to to hide the
implementation behind an interface.

**Comment by krousey** [comment]
> If StatusError had public members

It does [errors.StatusError](https://github.com/kubernetes/kubernetes/blob/master/pkg/api/errors/errors.go#L44). Only one public member:

```go
type StatusError struct {
	ErrStatus unversioned.Status
}
```

> and could be constructed by any package

and packages already do: [pkg/util/httpstream/spdy/roundtripper.go#L217](https://github.com/kubernetes/kubernetes/blob/master/pkg/util/httpstream/spdy/roundtripper.go#L217). This does a struct literal, but you can also use one of the many helper methods like `errors.NewUnauthorized`, `errors.NewForbidden`, and `errors.NewBadRequest`.


My main goal is to get APIStatus moved, so I will go ahead with number 2. Thanks for the feedback and being patient with this. :smile:

**Comment by krousey** [comment]
@lavalamp Thanks for the input. We should definitely work towards being able to separate the API as much as possible, and maybe having a separate errors package for the client makes sense down the road. I won't attempt to tackle this now as I think @smarterclayton and I have agreed on a quick fix to unblock more immediate needs.

**Comment by krousey** [comment]
I've rebased and changed the lines the referenced client.APIStatus to now reference errors.APIStatus.

File: pkg/kubectl/cmd/util/helpers.go [hunk 1]
```diff
@@ -57,7 +56,7 @@ type debugError interface {
 // souce is the filename or URL to the template file(*.json or *.yaml), or stdin to use to handle the resource.
 func AddSourceToErr(verb string, source string, err error) error {
 	if source != "" {
-		if statusError, ok := err.(*errors.StatusError); ok {
+		if statusError, ok := err.(errors.APIStatus); ok {
```
Comment (krousey) [inline pkg/kubectl/cmd/util/helpers.go]: Note that this is still inconsistent with line 108 of this file, but it was like that before.

**Comment by krousey** [comment]
@smarterclayton ping? If you're too busy, I can assign a different reviewer.

**Comment by smarterclayton** [comment]
LGTM thanks

**Comment by krousey** [comment]
Just a squash and rebase, re-adding LGTM.

**Comment by krousey** [comment]
@k8s-bot unit test this
